### DISCUSSION TITLE
Tests hards con código base sin prints se demoran mucho en WSL

### DISCUSSION THREAD

**Discussion by tatanas** [IIC2133-PUC/2024-2#171] [Tarea 3] [answered]
Hola hola! estaba probando los testcases y me di cuenta que algunos de la parte 1 ya se demoran 30 segundos con solamente el código base.
Acá están mis pruebas. Tomé el código base y le quité los prints a consola que ralentizaban más aún el programa. Por lo tanto, todo lo que se está ejecutando es el fscanf para cada uno de los ints entregados. Estoy usando un i7-10700F y con 16 GB de RAM.
![image](https://github.com/user-attachments/assets/f75f411c-d0b7-4905-8c7b-c949d55f9c1e)

Probé también los medium como sanity check.
Está bien que se demore tanto? Pueden probar ustedes, para ver si es un tema de mi computador que se esté demorando tanto? Si no, hay alguna otra solución para correr esto más rápido (online a lo mejor)?

Gracias!

**Comment by benjavicente** [comment, marked as answer]
Hola!

La solución rápida es mover los archivos fuera de`/mnt`:

```bash
mv -r ./tests ~/edd/T3/tests
./puffle ~/edd/T3/tests/ {remplazar por path al archivo}
```

---

El problema se ve en los tiempos que te entrega: tiempo real muy alto, pero tiempo de usuario bajo. **El tiempo real es el tiempo que uno ve, mientras que el tiempo de usuario es el tiempo CPU que ocupa el programa**.

Entonces, quiere decir que hay cosas externas a tu programa (llamadas al sistema operativo) que lo hacen lento, y que sin esos llamados puede bajar a los 2s que sale en el output que compartes. Como solo se lee archivos, el problema está en como se leen (donde ese como no tiene que ver con tu programa). Esto es algo que se ve en detalle en el curso de sistemas operativos, por si te interesa más.

Ahora, sobre la causa, es dado que en ese caso estás leyendo un archivo que está en otro sistema operativo. Dentro de Linux (WSL), estás leyendo un archivo de Windows (`/mnt/c`, montado al drive de C de Windows). Esos sistemas operativos no tienen el mismo sistema de archivos, entonces se debe realizar una conversión cada vez que se lee, que en WSL v2 es _muy_ lenta. [Esta pregunta en Stack Overflow sale de un setup más de web](https://stackoverflow.com/questions/68972448/why-is-wsl-extremely-slow-when-compared-with-native-windows-npm-yarn-processing), pero es el mismo problema.

La solución es mover los archivos que uno quiere usar en Linux hacia el sistema de archivos de Linux. Lo recomendado es que uno aprenda a [usar el terminal](https://www.cs.cornell.edu/courses/cs3410/2024fa/rsrc/shell.html) y trabajar completamente dentro de WSL (fuera de `/mnt/`) para que todo corra más rápido. Es decir, clonar el repositorio dentro de Linux y trabajar ahi. Windows igual [da la opción de ver los archivos con la UI de Explorer](https://learn.microsoft.com/en-us/windows/wsl/filesystems#view-your-current-directory-in-windows-file-explorer) (las primeras 3 sesiones de ese link las recomiendo leer).

Como este problema es más de setup, no afecta en como se corren los tests al evaluar.
Igual es clave aprender a evitarse esos problemas de WSL.
